File: izlaganja/predavanja/2007-08/prezentacija-20_prilozi/Java/tnserver.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Date;


public class tnserver {

	public static void main(String[] args) {
		
		if( args.length != 1 ) {
			System.out.println("usage: java tnserver <port>");
			System.exit(1);
		}
		
		ServerSocket ssocket = null;
			
		try {
			ssocket = new ServerSocket(Integer.parseInt(args[0]), 2);
		} catch (IOException e) {
			System.out.println("Error creating a server socket at port " + args[0]);
			System.exit(1);
		}
		
		System.out.println("Ready to accept connections at port " + args[0]);
		
		while(true) {
			
			Socket cSocket = null;
			PrintWriter out = null;
			BufferedReader in = null;
			
			try {
				cSocket = ssocket.accept();
				out = new PrintWriter(cSocket.getOutputStream());
				in = new BufferedReader(new InputStreamReader(cSocket.getInputStream()));
			} catch (IOException e) {
				System.out.println("Error in accepting the socket connection");
				System.exit(1);
			}
			
			String clientInfo = cSocket.getInetAddress().getHostName();
			System.out.println("Accepted connection from " +  clientInfo + ".");
			
			while(true) {
			
				try {
					String cmd = in.readLine();
				
					if( cmd.equalsIgnoreCase("time") ) {

						out.println("" + new Date());
						out.flush();
					}
					else if( cmd.equalsIgnoreCase("close") ) {
						break;
					}

					else {
						out.println("not understood: " + cmd);
						out.flush();
					}
				} catch(IOException e) {
					System.out.println("Communication error");
					break;
				}
			}
			
			
			System.out.println("Closing connection with client " + clientInfo + ".");
			
			out.close();
			try {
				in.close();
			} catch (IOException e) {
			}
			try {
				cSocket.close();
			} catch (IOException e) {
			}
			
		}
	}
}

File: izlaganja/predavanja/2007-08/prezentacija-20_prilozi/PHP/tnserver.php
<?php

	if( $argc != 2 ) {
		echo "usage: php tnserver.php <port>\n";
		echo "port between 1024 and 65535\n";
		exit(1);
	}
	
	$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
	if( $socket == FALSE ) reportError();
	if( socket_bind($socket, '127.0.0.1', $argv[1]) == FALSE ) reportError();
	if( socket_listen($socket, 2) == FALSE )  reportError();
	
	echo "Ready to accept connections at port $argv[1]\n";
	
	//main client loop
	while($cSocket = socket_accept($socket)) {
	
		$peerName = "";
		socket_getpeername($cSocket, $peerName);
		echo "Accepted connection from $peerName.\n";
		
		
		//client conversation loop
		while(TRUE) {
		
			$cmd = "";
			do {
				$cmf = @socket_read($cSocket, 1500);
				if($cmf == FALSE) break;
				$cmd .= $cmf;
			} while( $cmf[strlen($cmf)-1] != "\n");
	
			$cmd = stripCtrl($cmd);
			
			if( strcasecmp($cmd, 'TIME') == 0 ) {
				$status = socket_write($cSocket, date('n/j/Y g:i a') . "\n");
				if($status == FALSE) {
					socket_close($cSocket);
					echo "Connection with $peerName has been lost.\n";
					break;
				}
			}
			elseif( strcasecmp($cmd, 'CLOSE') == 0 ) {
			
				socket_write($cSocket, "server: closing connection\n");
				socket_close($cSocket);
				echo "Connection with $peerName closed.\n";
				break;
			}
			else {
				$status = @socket_write($cSocket, "not understood: $cmd\n");
				if($status == FALSE) {
					@socket_close($cSocket);
					echo "Connection with $peerName has been lost.\n";
					break;
				}
			}
		}
	}

	socket_close($socket);


	function stripCtrl($string) {
	
		$chars = str_split($string);
		$end = 0;
		while( ($end < strlen($string)) && (ord($chars[$end]) >= 0x30) ) $end++;
		
		return substr($string, 0, $end);
	}
	
	function reportError() {
	
		echo "(" . socket_last_error() . ") " . socket_strerror(socket_last_error()). "\n";
		exit(1);
	}
?>